### PR TITLE
State client

### PR DESCRIPTION
**Pull request by matis755**

### CODE REVIEW

**Comment by matis755** [comment]
Cppcheck fails with PRs, however after push, everything was fine with cppcheck

File: src/state_client.h [hunk 1]
```diff
@@ -0,0 +1,120 @@
+#ifndef AUTOSAR_STATE_MANAGEMENT_STATE_CLIENT_H
+#define AUTOSAR_STATE_MANAGEMENT_STATE_CLIENT_H
+/** @file state_client.h
+ * @brief State Client EM interface. */
+
+#include "sm_types.h"
+
+namespace ara { namespace exec {
```
Comment (r4jmond) [inline src/state_client.h]: please merge namespaces as we are using cpp17

File: tests/test_state_client.cpp [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        MOCK_METHOD(void, ExecSetState, (sm::FunctionGroupStateType));
+        MOCK_METHOD(void, FwSetState, (sm::FunctionGroupStateType));
+        MOCK_METHOD(void, IamSetState, (sm::FunctionGroupStateType));
+        MOCK_METHOD(void, IdsmSetState, (sm::FunctionGroupStateType));
+        MOCK_METHOD(void, LogSetState, (sm::FunctionGroupStateType));
+        MOCK_METHOD(void, NmSetState, (sm::FunctionGroupStateType));
+        MOCK_METHOD(void, NaSetState, (sm::FunctionGroupStateType));
+        MOCK_METHOD(void, PerSetState, (sm::FunctionGroupStateType));
+        MOCK_METHOD(void, PhmSetState, (sm::FunctionGroupStateType));
+        MOCK_METHOD(void, SmSetState, (sm::FunctionGroupStateType));
```
Comment (r4jmond) [inline tests/test_state_client.cpp]: SmSetState shall be real, but we can prepare state machine for SM later

File: tests/test_state_client.cpp [hunk 3]
```diff
@@ -0,0 +1,34 @@
+#include "state_client.h"
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+namespace ara { namespace exec {
```
Comment (r4jmond) [inline tests/test_state_client.cpp]: merge namespaces

**Review by r4jmond — COMMENTED** [review]
merge namespace and then this PR will be merged

File: src/state_client.h (hunk 1 above)
Comment (matis755) [reply src/state_client.h]: Done

File: tests/test_state_client.cpp (hunk 2 above)
Comment (matis755) [reply tests/test_state_client.cpp]: Done

File: tests/test_state_client.cpp (hunk 3 above)
Comment (matis755) [reply tests/test_state_client.cpp]: Done

File: src/state_management.h [hunk 4]
```diff
@@ -25,6 +26,8 @@ namespace ara { namespace sm {
         /** @brief Fills [SWS_SM_00021] */
         com::TriggerIn triggerIn;
         com::TriggerInOut triggerInOut;
+        SMStateType internalState = SMStateType::Off;
+        exec::StateClient * stateClient = nullptr;
```
Comment (r4jmond) [inline src/state_management.h]: shall be 'exec::StateClient* stateClient' or 'exec::StateClient *stateClient'. Nowadays, I think the first option is common

File: src/state_management.h [hunk 5]
```diff
@@ -9,13 +9,14 @@
...
 #include "network_handle.h"
+#include "state_client.h"
 
 namespace ara { namespace sm {
     /** @brief Class StateManagement to be used by Update and Configuration Management.
      *         Fills [SWS_SM_XXXXX]. */
     class StateManagement {
     public:
-        StateManagement();
+        StateManagement(exec::StateClient * sc);
```
Comment (r4jmond) [inline src/state_management.h]: shall be 'exec::StateClient* sc' or 'exec::StateClient *sc'. Nowadays, I think the first option is common

File: src/state_management.cpp [hunk 6]
```diff
@@ -18,12 +18,19 @@ namespace ara { namespace sm {
...
+        }
     }
 
     void StateManagement::Kill() {
         killFlag = true;
     }
 
-    StateManagement::StateManagement() :
-        myUpdateRequest{com::UpdateRequest()}, myNetworkHandle{com::NetworkHandle()}, killFlag{false} {}
+    StateManagement::StateManagement(exec::StateClient * sc) :
```
Comment (r4jmond) [inline src/state_management.cpp]: shall be 'exec::StateClient* stateClient' or 'exec::StateClient *stateClient'. Nowadays, I think the first option is common

**Review by r4jmond — COMMENTED** [review]
just some minor thing

**Review by r4jmond — COMMENTED** [review]
move pointer star

File: src/state_management.h (hunk 4 above)
Comment (matis755) [reply src/state_management.h]: Done

File: src/state_management.h (hunk 5 above)
Comment (matis755) [reply src/state_management.h]: Done

File: src/state_management.cpp (hunk 6 above)
Comment (matis755) [reply src/state_management.cpp]: Done

**Review by r4jmond — APPROVED** [review]

**Pull request merged**